### PR TITLE
Remove chunker.Reset() and chunker pool

### PR DESCRIPTION
**Pull request by fw42**
This removes pools.go and makes the Chunker use a simpler pool for it's buffer. Also make `chunker.Reset()` private (since we are not reusing chunker instances anymore).

### CODE REVIEW

**Comment by fw42** [comment]
As discussed in https://github.com/restic/restic/pull/166#discussion_r29563676, I would like `chunker.New()` to return an error if the given polynomial is not irreducible. Can be a follow-up PR though.

**Comment by fd0** [comment]
benchcmp:

    benchmark                         old ns/op      new ns/op      delta
    BenchmarkNodeFillUser             1028           941            -8.46%
    BenchmarkNodeFromFileInfo         1521           1154           -24.13%
    BenchmarkChunkEncrypt             195677058      194909636      -0.39%
    BenchmarkChunkEncryptParallel     154972703      151944471      -1.95%
    BenchmarkArchiveDirectory         1644442211     1789486599     +8.82%
    BenchmarkLoadTree                 1524768        1286182        -15.65%
    BenchmarkChunkerWithSHA256        149588556      149241978      -0.23%
    BenchmarkChunkerWithMD5           94464237       149997764      +58.79%
    BenchmarkChunker                  74336464       149365064      +100.93%
    BenchmarkNewChunker               74268          96291          +29.65%
    BenchmarkPolDivMod                69.7           68.4           -1.87%
    BenchmarkPolDiv                   70.1           73.2           +4.42%
    BenchmarkPolMod                   70.0           73.1           +4.43%
    BenchmarkPolDeg                   44.3           44.2           -0.23%
    BenchmarkRandomPolynomial         390182169      338223606      -13.32%
    BenchmarkEncryptWriter            37556242       37103392       -1.21%
    BenchmarkEncrypt                  37154428       37048124       -0.29%
    BenchmarkDecryptReader            37709789       37362253       -0.92%
    BenchmarkEncryptDecryptReader     76913199       76057854       -1.11%
    BenchmarkDecrypt                  35646165       35979590       +0.94%
    BenchmarkPipelineWalker           377325         372170         -1.37%
    BenchmarkSaveJSON                 18681          18682          +0.01%
    BenchmarkSaveFrom                 55962017       56514071       +0.99%

    benchmark                         old MB/s     new MB/s     speedup
    BenchmarkChunkEncrypt             53.59        53.80        1.00x
    BenchmarkChunkEncryptParallel     67.66        69.01        1.02x
    BenchmarkChunkerWithSHA256        70.10        70.26        1.00x
    BenchmarkChunkerWithMD5           111.00       69.91        0.63x
    BenchmarkChunker                  141.06       70.20        0.50x
    BenchmarkEncryptWriter            223.36       226.09       1.01x
    BenchmarkEncrypt                  225.78       226.42       1.00x
    BenchmarkDecryptReader            222.45       224.52       1.01x
    BenchmarkEncryptDecryptReader     109.07       110.29       1.01x
    BenchmarkDecrypt                  235.33       233.15       0.99x
    BenchmarkSaveFrom                 74.95        74.22        0.99x

    benchmark                         old allocs     new allocs     delta
    BenchmarkNodeFillUser             6              6              +0.00%
    BenchmarkNodeFromFileInfo         2              2              +0.00%
    BenchmarkChunkEncrypt             110            113            +2.73%
    BenchmarkChunkEncryptParallel     28             31             +10.71%
    BenchmarkArchiveDirectory         197556         207626         +5.10%
    BenchmarkLoadTree                 1675           1444           -13.79%
    BenchmarkChunkerWithSHA256        13             16             +23.08%
    BenchmarkChunkerWithMD5           12             16             +33.33%
    BenchmarkChunker                  6              16             +166.67%
    BenchmarkNewChunker               3              4              +33.33%
    BenchmarkPolDivMod                0              0              +0.00%
    BenchmarkPolDiv                   0              0              +0.00%
    BenchmarkPolMod                   0              0              +0.00%
    BenchmarkPolDeg                   0              0              +0.00%
    BenchmarkRandomPolynomial         99             44             -55.56%
    BenchmarkEncryptWriter            16             16             +0.00%
    BenchmarkEncrypt                  12             12             +0.00%
    BenchmarkDecryptReader            14             14             +0.00%
    BenchmarkEncryptDecryptReader     29             29             +0.00%
    BenchmarkDecrypt                  10             10             +0.00%
    BenchmarkPipelineWalker           197            198            +0.51%
    BenchmarkSaveJSON                 59             59             +0.00%
    BenchmarkSaveFrom                 189            188            -0.53%

    benchmark                         old bytes     new bytes     delta
    BenchmarkNodeFillUser             80            80            +0.00%
    BenchmarkNodeFromFileInfo         356           356           +0.00%
    BenchmarkChunkEncrypt             64705         536862        +729.71%
    BenchmarkChunkEncryptParallel     55083         527240        +857.17%
    BenchmarkArchiveDirectory         105524756     748466208     +609.28%
    BenchmarkLoadTree                 123895        100943        -18.53%
    BenchmarkChunkerWithSHA256        581           525244        +90303.44%
    BenchmarkChunkerWithMD5           434           525244        +120923.96%
    BenchmarkChunker                  338           525244        +155297.63%
    BenchmarkNewChunker               262512        524784        +99.91%
    BenchmarkPolDivMod                0             0             +0.00%
    BenchmarkPolDiv                   0             0             +0.00%
    BenchmarkPolMod                   0             0             +0.00%
    BenchmarkPolDeg                   0             0             +0.00%
    BenchmarkRandomPolynomial         2112          938           -55.59%
    BenchmarkEncryptWriter            282733        170331        -39.76%
    BenchmarkEncrypt                  1600          1600          +0.00%
    BenchmarkDecryptReader            569243        569243        +0.00%
    BenchmarkEncryptDecryptReader     844868        844866        -0.00%
    BenchmarkDecrypt                  1573          1573          +0.00%
    BenchmarkPipelineWalker           57944         57995         +0.09%
    BenchmarkSaveJSON                 2907          2922          +0.52%
    BenchmarkSaveFrom                 25211071      25210789      -0.00%

File: chunker/chunker.go [hunk 1]
```diff
@@ -17,6 +17,9 @@ const (
 	// aim to create chunks of 20 bits or about 1MiB on average.
 	averageBits = 20
 
+	// default buffer size
+	bufSize = 512 * KiB
```
Comment (fw42) [inline chunker/chunker.go]: I guess this is not needed anymore now?

**Comment by fw42** [comment]
:+1:

File: archiver.go [hunk 2]
```diff
@@ -169,6 +170,12 @@ func updateNodeContent(node *Node, results []saveResult) error {
 	return nil
 }
 
+const chunkerBufSize = 512 * chunker.KiB
+
+var chunkerBufPool = sync.Pool{
```
Comment (fw42) [inline archiver.go]: as discussed in IRC, maybe this should live on the chunker type as a non-exported field?

**Comment by fd0** [comment]
Interesting benchmarks:

    % benchcmp /tmp/{before,after}
    benchmark                         old ns/op      new ns/op      delta
    [...]
    BenchmarkArchiveDirectory         1863372190     1052511209     -43.52%
    BenchmarkLoadTree                 2108148        1419818        -32.65%
    BenchmarkChunkerWithSHA256        148980904      150942129      +1.32%
    BenchmarkChunkerWithMD5           94357514       95597196       +1.31%
    BenchmarkChunker                  74275711       75383805       +1.49%
    BenchmarkNewChunker               74945          607            -99.19%
    [...]

    benchmark                         old bytes     new bytes     delta
    [...]
    BenchmarkChunkEncrypt             64694         12556         -80.59%
    BenchmarkChunkEncryptParallel     55072         2940          -94.66%
    BenchmarkArchiveDirectory         130384200     130690536     +0.23%
    BenchmarkLoadTree                 165816        115941        -30.08%
    BenchmarkChunkerWithSHA256        581           828           +42.51%
    BenchmarkChunkerWithMD5           434           678           +56.22%
    BenchmarkChunker                  338           574           +69.82%
    BenchmarkNewChunker               262512        336           -99.87%
    [...]

So we didn't break anything serious. :+1:

File: archiver_test.go [hunk 3]
```diff
@@ -17,7 +18,7 @@ import (
 var benchArchiveDirectory = flag.String("test.benchdir", ".", "benchmark archiving a real directory (default: .)")
 var testPol = chunker.Pol(0x3DA3358B4DC173)
 
-const bufSize = chunker.MiB
+const chunkerBufSize = 512 * chunker.KiB
```
Comment (fw42) [inline archiver_test.go]: this can be removed now

File: chunker/chunker.go [hunk 4]
```diff
@@ -79,38 +85,36 @@ type Chunker struct {
 	h      hash.Hash
 }
 
-// New returns a new Chunker based on polynomial p that reads from data from rd
-// with bufsize and pass all data to hash along the way.
-func New(rd io.Reader, p Pol, bufsize int, hash hash.Hash) *Chunker {
+// New returns a new Chunker based on polynomial p that reads from rd
+// with bufsize and pass all data to hash along the way, using buf for
```
Comment (fw42) [inline chunker/chunker.go]: > using buf for ...

can be removed now

File: chunker/chunker.go [hunk 5]
```diff
@@ -200,6 +204,9 @@ func (c *Chunker) Next() (*Chunk, error) {
 			if err == io.EOF && !c.closed {
 				c.closed = true
 
+				// return the buffer to the pool
+				// bufPool.Put(c.buf)
```
Comment (fw42) [inline chunker/chunker.go]: what's this? remove?

File: chunker/chunker_test.go [hunk 6]
```diff
@@ -148,10 +147,12 @@ func getRandom(seed, count int) []byte {
 	return buf
 }
 
+const chunkerBufSize = 512 * chunker.KiB
```
Comment (fw42) [inline chunker/chunker_test.go]: can be removed

**Pull request merged**